nexus/src/app/disk.rs

```diff
@@ -408,4 +352,277 @@ impl super::Nexus {
+            let data: Vec<u8> = base64::Engine::decode(
+                &base64::engine::general_purpose::STANDARD,
+                &param.base64_encoded_data,
+            )
+            .map_err(|e| {
+                Error::invalid_request(&format!(
+                    "error base64 decoding data: {}",
+                    e
+                ))
+            })?;
+
+            info!(
+                self.log,
+                "bulk write of {} bytes to offset {} of disk {} using pantry endpoint {:?}",
+                data.len(),
+                param.offset,
+                db_disk.id(),
+                endpoint,
+            );
+
+            // The the disk state can change between the check above and here
+            // because there's no disk state change associated with this write.
+            // I believe that toggling the disk's state and generation number in
+            // the DB is too expensive to be done for every 512k chunk of a disk
+            // (or whatever the eventual import chunk size is), however I didn't
+            // actually measure it.
+            //
+            // For example, between the check against state
+            // ImportingFromBulkWrites and here, the user could have called the
+            // bulk-write-stop endpoint, which would put the disk into state
+            // ImportReady. They could have then called the finalize endpoint,
+            // which would kick off the disk finalizing saga: set the state to
+            // finalzing, optionally take a snapshot, detach it from the
+            // associated Pantry, and set the state to Detached.
+            //
+            // In this scenario the write here would fail because the volume
+            // would have been detached from the Pantry, but say that the user
+            // instead called bulk-write-stop, then import, thereby kicking off
+            // the import blocks from URL saga? The Pantry locks its internal
+            // entry by the ID of the volume, so at least the requests would be
+            // ordered by their arrival time: either this bulk write would land
+            // first, or the import would land first:
+            //
+            // - if this bulk write landed first, the import would likely
+            //   overwrite it with blocks from a URL.
+            //
+            // - if the import landed first, then the bulk write would overwrite
+            //   what the import job imported, probably corrupting the data.
+            //
+            // I also believe it's not correct to use a saga here. Really, the
+            // user's cli (or any other program really) is responsible for this
+            // bulk write operation, not Nexus. If the bulk_write call below
+            // fails, then that failure would be propagated up to the user, and
+            // that user's program can act accordingly. In a way, the user's
+            // program is an externally driven saga instead.
+
```

Comment:
That's right, yeah. https://github.com/oxidecomputer/crucible/pull/654 will add a validate command to the Pantry, allowing the user to checksum the whole disk and compare it against their known good one. Once that lands I'll update Nexus' API so that a checksum can be passed during the finalize step.